op-tests/test-bls-ops.txt

```diff
@@ -22,6 +22,117 @@ coinid 0x1234500000000000000000000000000000000000000000000000000000000000 0x6789
 coinid 0x1234500000000000000000000000000000000000000000000000000000000000 0x6789abcdef000000000000000000000000000000000000000000000000000000 -1 => FAIL
 coinid 0x1234500000000000000000000000000000000000000000000000000000000000 0x6789abcdef000000000000000000000000000000000000000000000000000000 0x00001234 => FAIL
 
+; modpow
+modpow => FAIL
+modpow 1 => FAIL
+modpow 1 2 => FAIL
+modpow 1 2 3 4 => FAIL
+modpow (1 2) 2 3 => FAIL
+modpow 1 (2 2) 3 => FAIL
+modpow 1 2 (2 3) => FAIL
+
+; raised to negative
+modpow 999 -1 10 => FAIL
+modpow 2 -1 10 => FAIL
+
+; mod of 0
+modpow 2 2 0 => FAIL
+modpow 2 3 0 => FAIL
+
+modpow 2 0 100 => 1 | 11033
+modpow 2 1 100 => 2 | 11036
+modpow 2 2 100 => 4 | 11036
+modpow 2 3 100 => 8 | 11036
+
+modpow 2 0 -10 => -9 | 11033
+modpow 2 1 -10 => -8 | 11036
+modpow 2 2 -10 => -6 | 11036
+modpow 2 3 -10 => -2 | 11036
+
+modpow 2 0 10 => 1 | 11033
+modpow 999 0 10 => 1 | 11048
+
+modpow 12345 6789 4444 => 2105 | 11094
+modpow -12345 6789 4444 => 2339 | 11094
+modpow -12345 6789 -4444 => -2105 | 11094
+modpow 12345 6789 -4444 => -2339 | 11094
+
+modpow 12345 6789 44444444444 => 13456191581 | 11292
+
+modpow 0 6789 4444 => 0 | 11044
+modpow 2 64 32 => 0 | 11026
+modpow 2 64 31 => 16 | 11036
+
```

Comment:
A typical use case for secp would be three random-looking 256 bit numbers. Maybe add a couple tests that look like that?

Reply:
These costs seem way too low. Have you benchmarked this? I'd think the cost should scale linearly with the log of the exponent, but also with the size of the modulus, since the implementation is probably square $log_2 X$ times (where X = exponent) and multiply the accumulated result about half the time.

Reply:
Maybe there's a smarter implementation I'm not aware of.

Reply:
It uses Montgomery multiplication in the case of an odd modulus. I thought if the modulus were even, it would do some reduction to make the modulus odd, but I don't see any evidence of that. So maybe this is a really smart and really fast implementation, at least for odd modulus. For secp, the modulus is constant, and odd.

Reply:
https://github.com/rust-num/num-bigint/blob/master/src/biguint/monty.rs

Reply:
I see you've added additional tests in a later change.

Reply:
I did benchmark the cost, but I did not perform an exhaustive search, and I made the assumption/simplification that the cost grows exponentially in the size of the exponent and the modulus. Specifically, I just searched one parameter at a time, leaving the others with relatively small numbers. At small numbers the cost is probably dominated by the parsing and allocation of the number.

I'll add some rigour to these measurements and post an update

Reply:
I just noticed I also had failed to add the python program I used to generate the test cases. This is what validates the results against python. I'll add that as adjust the test cases.